### PR TITLE
Convert TextFieldMapper to parametrized form

### PR DESCRIPTION
**Pull request by romseygeek**
As a result of this, we can remove a chunk of code from `TypeParsers` as well.  Tests
for search/index mode analyzers have moved into their own file.  This commit also
rationalises the serialization checks for parameters into a single `SerializerCheck`
interface that takes the values `includeDefaults`, `isConfigured` and the value
itself.

Relates to #62988

### CODE REVIEW

File: plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java [hunk 1]
```diff
@@ -72,92 +70,94 @@
...
+        private final Parameter<Boolean> index = Parameter.indexParam(m -> builder(m).index.getValue(), true);
+        private final Parameter<Boolean> store = Parameter.storeParam(m -> builder(m).store.getValue(), false);
 
-        public Builder positionIncrementGap(int positionIncrementGap) {
-            if (positionIncrementGap < 0) {
-                throw new MapperParsingException("[positions_increment_gap] must be positive, got " + positionIncrementGap);
-            }
-            this.positionIncrementGap = positionIncrementGap;
-            return this;
+        final TextParams.Analyzers analyzers;
```
Comment (nik9000) [inline plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java:87]: Could you move this below all the `Parameter` so I don't lose it?

File: plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java [hunk 2]
```diff
@@ -191,22 +191,18 @@ public static AnnotatedText parse (String textPlusMarkup) {
...
-                        if (value!=null && value.length() > 0) {
-                            annotations.add(new AnnotationToken(startOffset, endOffset, value));
+                    if(kv.length == 2){
+                        throw new ElasticsearchParseException("key=value pairs are not supported in annotations");
+                    }
+                    if(kv.length == 1) {
+                        //Check "=" sign wasn't in the pair string
+                        if(kv[0].length() == pair.length()) {
+                            //untyped value
+                            value = URLDecoder.decode(kv[0], StandardCharsets.UTF_8);
```
Comment (nik9000) [inline plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java:201]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java [hunk 3]
```diff
@@ -142,7 +142,7 @@ private static CompletionFieldMapper toType(FieldMapper in) {
             m -> toType(m).maxInputLength, Defaults.DEFAULT_MAX_INPUT_LENGTH)
             .addDeprecatedName("max_input_len")
             .setValidator(Builder::validateInputLength)
-            .alwaysSerialize();
+            .setSerializerCheck((id, ic, v) -> true);
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java:145]: I wonder if `alwaysSerialize` should a shortcut to this new thing - it is easier for me to read the intent of `alwaysSerialize` than this lambda.

File: server/src/main/java/org/elasticsearch/index/mapper/ParametrizedFieldMapper.java [hunk 4]
```diff
@@ -178,6 +180,11 @@ public T getValue() {
             return isSet ? value : defaultValue.get();
         }
 
+        @Override
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/ParametrizedFieldMapper.java:183]: Are you thinking of removing the old `getValue` in a follow up?

File: server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java [hunk 5]
```diff
@@ -121,101 +119,230 @@
...
+        Map<?, ?> indexPrefix = (Map<?, ?>) propNode;
+        int minChars = XContentMapValues.nodeIntegerValue(indexPrefix.remove("min_chars"),
+            Defaults.INDEX_PREFIX_MIN_CHARS);
+        int maxChars = XContentMapValues.nodeIntegerValue(indexPrefix.remove("max_chars"),
+            Defaults.INDEX_PREFIX_MAX_CHARS);
+        DocumentMapperParser.checkNoRemainingFields(propName, indexPrefix, parserContext.indexVersionCreated());
+        return new PrefixConfig(minChars, maxChars);
+    }
+
+    private static final class FielddataFrequencyFilter implements ToXContent {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:186]: :+1:

**Review by nik9000 — APPROVED** [review]
Everything seems in order. I've not looked as closely as I maybe should - I'm trusting @romseygeek and tests to cover sneaky stuff. The less obvious changes that I saw all make good sense to me.

File: server/src/main/java/org/elasticsearch/index/mapper/ParametrizedFieldMapper.java (hunk 4 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/ParametrizedFieldMapper.java:183]: That's the plan, yes

**Comment by romseygeek** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java (hunk 3 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java]: I added `alwaysSerialize` and `neverSerialize` as shortcuts.

**Comment by romseygeek** [comment]
BWC issues with serialization mean that I've had to override `toXContentBody` and expose a bit more of the internal workings of Parameter than I'd like; I think if we remove the Mapper consistency assertions (see #59427) we can fix this up again in master, but for now we have this nasty sticking plaster.

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Pull request merged**